lib/core/src/Cardano/Wallet/StakePool/Metrics.hs

```
@@ -14,37 +16,169 @@ module Cardano.Wallet.StakePool.Metrics
+import Data.List.NonEmpty
+    ( NonEmpty )
+import Data.Map.Merge.Strict
+    ( WhenMissing, mergeA, traverseMissing, zipWithMatched )
 import Data.Map.Strict
     ( Map )
-import Data.Word
-    ( Word64 )
+import Data.Text
+    ( Text )
 import Fmt
     ( Buildable (..), blockListF', fmt, (+|), (|+) )
 import GHC.Generics
     ( Generic )
 
 import qualified Data.Map.Strict as Map
 
+--
+-- Metrics
+--
+
+data ErrListStakePools
+    = ErrListStakePoolsStakeIsUnreachable
+    | ErrListStakePoolsMetricsIsUnsynced
+    | ErrListStakePoolsInternalError ErrListStakePoolsInternalErr
+
+data ErrListStakePoolsInternalErr
+    = ErrListStakePoolInternalErrInconsistentData
+    | ErrListStakePoolInternalErrReadStateErr
+
+newtype StakePoolLayer m = StakePoolLayer
+    { listStakePools
+        :: ExceptT ErrListStakePools m [(PoolId, (Stake, Int))]
+    }
+
+newStakePoolLayer
+    :: NetworkLayer IO tx (BlockHeader, PoolId)
+    -> Trace IO Text
+    -> IO (StakePoolLayer IO)
+newStakePoolLayer nl tr = do
+    mvar <- worker nl tr
+    return $ StakePoolLayer
+        { listStakePools = Map.toList <$>
+            combineMetrics
+                (withE1 $ stakeDistribution nl)
+                (liftIO $ readMVar mvar)
+        }
+  where
+    withE1 = withExceptT $ const ()
+
+-- | Combines two different sources of data: stake distribution and pool activity
+-- together.
+combineMetrics
+    :: forall m readStateErr stakeDistrErr. Monad m
+    => ExceptT stakeDistrErr m (EpochNo, Map PoolId Stake)
+    -> ExceptT readStateErr m State
+    -> ExceptT ErrListStakePools m (Map PoolId (Stake, Int))
+combineMetrics getStakeDistr getActivityState = do
+    (epoch, distr) <- withExceptT (const ErrListStakePoolsStakeIsUnreachable)
+        getStakeDistr
+
+    s <- withInternalErr ErrListStakePoolInternalErrReadStateErr
+        getActivityState
+
+    m <- case activityForEpoch epoch s of
+        Just x -> return x
+        Nothing -> throwE ErrListStakePoolsMetricsIsUnsynced
```

What about returning an empty map? I don't think the error is needed.

---

That doesn't sound right to me. With an error, we can report the error like:

```
I can't list stake pools yet because I need to scan the blockchain for metrics first. I'm at 1%
```